torch/csrc/autograd/engine.cpp

```diff
@@ -402,4 +431,71 @@ auto Engine::start_threads() -> void {
   }
 }
 
+void GraphTask::init_to_execute(const std::shared_ptr<Function>& graph_root, const function_list& outputs) {
+  exec_info[graph_root.get()].needed = true;
```

Comment:
Do the "output" variables all point to the graph_root in their next_functions?

Reply:
not really. `graph_root->next_functions` are the roots of `execute`. `output`s are where we stop the execution